### ISSUE TITLE
Incompatibility with libgit2@1.8

### ISSUE THREAD

**Issue by meiyasan** [RogerGee/php-git2#4]
Hi Roger, 

I just noticed that php-git2 has issues with libgit2 > 1.7.2 (last version I tested)
I know php-git2 is officially supporting only up to 1.5, but I wanted to mention it for the record.

I tried to fix it, but i don't have good solution at the moment. 
Thank you again for the great work. It works really nicely with php8.2
Cheers, M.

**Comment by RogerGee** [comment]
@xkzl It looks like the latest releases of `libgit2`  (`1.8.0` and `1.8.1`) introduced a breaking change in the `git_commit` functions by changing the `const`-ness of a few parameters. It also looks like the maintainers are [reverting the changes](https://github.com/libgit2/libgit2/commit/49d3fadfca4ce8e7a643525eb301a2d45956641e) in the next release (i.e. `1.8.2`).

While this issue can be fixed with a few type casts, I'd like to wait and see how the `1.8.2` release turns out. If they keep the original declarations, then we don't have to fix that.

I also saw another error in the `php-config-backend.cpp` module:

```
 g++ -I. -I/home/roger/code/projects/php-git2 -I/home/roger/code/projects/php-git2/include -I/home/roger/code/projects/php-git2/main -I/home/roger/code/projects/php-git2 -I/opt/rogerbox/php8.1-debug/include/php -I/opt/rogerbox/php8.1-debug/include/php/main -I/opt/rogerbox/php8.1-debug/include/php/TSRM -I/opt/rogerbox/php8.1-debug/include/php/Zend -I/opt/rogerbox/php8.1-debug/include/php/ext -I/opt/rogerbox/php8.1-debug/include/php/ext/date/lib -I/opt/rogerbox/libgit2-1.8.1-debug/include -DHAVE_CONFIG_H -g -O0 -Wfatal-errors -Wall -Wno-deprecated-register -std=c++11 -c /home/roger/code/projects/php-git2/php-git2-fe.cpp  -fPIC -DPIC -o .libs/php-git2-fe.o
In file included from /opt/rogerbox/php8.1-debug/include/php/Zend/zend.h:30,
                 from /opt/rogerbox/php8.1-debug/include/php/main/php.h:31,
                 from /home/roger/code/projects/php-git2/php-git2.h:15,
                 from /home/roger/code/projects/php-git2/php-array.h:9,
                 from /home/roger/code/projects/php-git2/php-type.h:9,
                 from /home/roger/code/projects/php-git2/php-object.h:9,
                 from /home/roger/code/projects/php-git2/php-config-backend.cpp:10:
/home/roger/code/projects/php-git2/php-config-backend.cpp: In function ‘void custom_backend_entry_free(git_config_entry*)’:
/home/roger/code/projects/php-git2/php-config-backend.cpp:38:16: error: ‘git_config_entry’ {aka ‘struct git_config_entry’} has no member named ‘payload’
   38 |     efree(ent->payload);
      |                ^~~~~~~
/opt/rogerbox/php8.1-debug/include/php/Zend/zend_alloc.h:161:81: note: in definition of macro ‘efree’
  161 | #define efree(ptr)                                                      _efree((ptr) ZEND_FILE_LINE_CC ZEND_FILE_LINE_EMPTY_CC)
      |                                                                                 ^~~
```

This is caused by the maintainers of `libgit2` dropping the `payload` field from the `git_config_entry` struct. `php-git2` took advantage of this structure to reference the base buffer pointer for when the custom config entry object is freed. I will need to refactor this code to allocate a custom substructure that allocates an additional field. (I could also just free `ent->name` since it is the same pointer.)

I will push this fix on a maintenance branch sometime later this week when I get a chance. I'll follow up here when I have it. In the meantime, you can apply this patch, which *should* work:

~~~
iff --git a/php-config-backend.cpp b/php-config-backend.cpp
index 5f61a58..570960a 100644
--- a/php-config-backend.cpp
+++ b/php-config-backend.cpp
@@ -35,7 +35,7 @@ static int config_backend_has_property(
 
 static void custom_backend_entry_free(git_config_entry* ent)
 {
-    efree(ent->payload);
+    efree(const_cast<char*>(ent->name));
     efree(ent);
 }
 
@@ -101,7 +101,7 @@ static int set_custom_backend_entry(zval* arr,git_config_entry* ent,const char**
 
     ent->level = level;
     ent->free = custom_backend_entry_free;
-    ent->payload = buffer;
+    //ent->payload = buffer;
 
     if (sname != nullptr) {
         zend_string_release(sname);
~~~

(I'm guessing you aren't using custom config backends, which makes this working a moot point. You really just need it to compile!)

**Comment by RogerGee** [comment]
@xkzl  Please check out the `php8/maint` branch. I've pushed a few commits to fix the 1.8 compatibility issues we noticed.

If the maintainers of libgit2 decide to revert the `git_commit* const[]` changes for the next 1.8 release, then we will have to revert also. I'm going to hold off making a release until we see what the next release does. Also, there are a few other things I'd like to check on regarding 1.5 => 1.8.

I'm going to go ahead and close this issue since you can work off the maintenance branch. But expect to see new releases hopefully soon. I'll also try to work on a new minor version release that implements additional bindings. (I'm sure recent versions of libgit2 have added a lot.)
